File: src/ParallelPatterns/src/Common/Helpers/ImageHandler.cs
using System;
using System.Drawing;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Formats.Jpeg;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;
using SixLabors.ImageSharp.Processing.Processors;
using SixLabors.ImageSharp.Processing.Processors.Overlays;
using SixLabors.Primitives;

namespace Helpers
{
    public static  class ImageHandler
    {
        private static readonly Configuration configuration = new Configuration(new JpegConfigurationModule());
        public static Image<Rgba32> Resize(Image<Rgba32> source, int newWidth, int newHeight)
        {
            // this.configuration.MaxDegreeOfParallelism = Environment.ProcessorCount;
            var image = source.Clone();
            image.Mutate(x => x.Resize(400, 400));
            return image;
        }

        public static Image<Rgba32> ConvertTo3D(Image<Rgba32> source)
        {
            var image = source.Clone();
            var w = image.Width;
            var h = image.Height;
            for (int x = 20; x < w; x++)
            {
                for (int y = 0; y < h; y++)
                {
                    var c1 = image[x, y];
                    var c2 = image[x - 20, y];
                    image[x - 20, y] = new Rgba32(c1.R, c2.G, c2.B);
                }
            }
            return image;
        }


        public static Image<Rgba32> SetFilter(Image<Rgba32> source, ImageFilters imageFilters)
        {
            var image = source.Clone();
            var w = image.Width;
            var h = image.Height;
            for (int x = 0; x < w; x++)
            {
                for (int y = 0; y < h; y++)
                {
                    var p = image[x, y];
                    if (imageFilters == ImageFilters.Red)
                    {
                        p.B = 0;
                        p.G = 0;
                    }
                    else if(imageFilters == ImageFilters.Green)
                    {
                        p.R = 0;
                        p.B = 0;
                    }
                    else if (imageFilters == ImageFilters.Blue)
                    {
                        p.R = 0;
                        p.G = 0;
                    }
                    else if (imageFilters == ImageFilters.Gray)
                    {
                        var gray = (byte)((0.299 * (double)p.R) + (0.587 * (double)p.G) + (0.114 * (double)p.B));
                        p.R = gray;
                        p.G = gray;
                        p.B = gray;
                    }
                    image[x, y] = p; 
                }
            }
            return image;
        }
    }

    public enum ImageFilters
    {
        Red,Green,Blue,Gray
    }
}
        

File: src/ParallelPatterns/src/Common/Helpers.FSharp/ImageHandler.fs
namespace HelpersFSharp

module ImageHandler =

    open System
    open System.Drawing
    open System.IO
    open SixLabors.ImageSharp
    open SixLabors.ImageSharp.Formats.Jpeg
    open SixLabors.ImageSharp.PixelFormats
    open SixLabors.ImageSharp.Processing
    open SixLabors.ImageSharp.Processing.Processors
    open SixLabors.ImageSharp.Processing.Processors.Overlays
    open SixLabors.Primitives

    let combine (target:string, source1:string, source2:string) =
        let img1 = Image.Load(source1)
        let img2 = Image.Load(source2)
        let combo = img1.Clone()
        let pixelize (img:Image<Rgba32>) = [
            for x in 0..img.Width - 1 do
                for y in 0..img.Height - 1 do
                    yield (x,y, img.[x,y]) ]
        let pix1 = pixelize img1
        let pix2 = pixelize img2
        let sorter (_,_,c:Rgba32) = [c.R;c.G;c.B] |> Seq.max
        (pix1 |> List.sortBy sorter, pix2 |> List.sortBy sorter)
        ||> List.zip
        |> List.iter(fun ((x1,y1,_), (_,_,c2)) -> 
            combo.[x1,y1] <- c2)
        combo.Save(target)

